File: android/app/src/main/java/com/example/yugiohscanner/ml/SaleNotices.kt
package com.example.yugiohscanner.ml

/**
 * Spec H3b §7.6 (H3b2) -- eBay-Hinweise (sale_notices) und Marke „Gebühren vorläufig“ (ebay_orders). ZWILLING von
 * desktop/src/utils/saleNotices.js, gemeinsame Fixture docs/fixtures/ebay/notices.json. Wer eine Fassung ändert,
 * ändert beide. Namensabweichungen: Felder camelCase; JS-`orders[saleId]` = Map saleId -> [OrderMark].
 * Sortierung per Codeeinheiten (String.compareTo), kein Collator.
 */
object SaleNotices {
    data class Notice(
        val noticeId: String, val kind: String, val text: String, val saleId: String?, val listingId: String?,
        val dismissed: Boolean, val createdAt: String?,
    )
    data class Label(val label: String, val tone: String)
    data class OrderMark(val status: String, val feesFinal: Boolean)

    const val FEES_PROVISIONAL = "Gebühren vorläufig"

    private val LABELS = mapOf(
        "shipping" to Label("Versand", "warn"),
        "error" to Label("Problem", "bad"),
        "reminder" to Label("Erinnerung", "neutral"),
        "token" to Label("eBay-Verbindung", "bad"),
    )

    /** Offene Hinweise, neueste zuerst; gleiche Zeit nach notice_id. */
    fun sort(list: List<Notice>): List<Notice> =
        list.filter { !it.dismissed }.sortedWith { a, b ->
            val t = (b.createdAt ?: "").compareTo(a.createdAt ?: "")
            if (t != 0) t else a.noticeId.compareTo(b.noticeId)
        }

    fun label(kind: String): Label = LABELS[kind] ?: Label("Hinweis", "neutral")

    fun feesMark(orders: Map<String, OrderMark>, saleId: String): String? {
        val o = orders[saleId] ?: return null
        return if (o.status == "gebucht" && !o.feesFinal) FEES_PROVISIONAL else null
    }

    fun title(count: Int): String = "eBay-Hinweise ($count)"
}

File: desktop/src/utils/saleNotices.js
// Spec H3b §7.6 (H3b2) -- eBay-Hinweise (sale_notices) und Marke „Gebühren vorläufig“ (ebay_orders), ohne React.
// ZWILLING: android .../ml/SaleNotices.kt. Gemeinsame Fixture: docs/fixtures/ebay/notices.json. Wer eine Fassung ändert,
// ändert beide. Sortierung per Codeeinheiten (kein localeCompare).

const LABELS = {
  shipping: { label: 'Versand', tone: 'warn' },
  error: { label: 'Problem', tone: 'bad' },
  reminder: { label: 'Erinnerung', tone: 'neutral' },
  token: { label: 'eBay-Verbindung', tone: 'bad' },
};
const cmp = (a, b) => (a < b ? -1 : a > b ? 1 : 0);

// Offene Hinweise, neueste zuerst; gleiche Zeit nach notice_id. dismissed kommt lokal als 0/1, aus der Cloud als true/false.
export function sortNotices(list) {
  return (list || []).filter((n) => !n.dismissed)
    .sort((a, b) => cmp(b.created_at ?? '', a.created_at ?? '') || cmp(a.notice_id, b.notice_id));
}

export function noticeLabel(kind) {
  return LABELS[kind] ?? { label: 'Hinweis', tone: 'neutral' };
}

// orders: sale_id -> { status, fees_final } (IPC ebay-orders).
export function feesMark(orders, saleId) {
  const o = orders?.[saleId];
  return o && o.status === 'gebucht' && !o.fees_final ? 'Gebühren vorläufig' : null;
}

export function noticesTitle(count) {
  return `eBay-Hinweise (${count})`;
}
